File: src/runtime/dart/built_in/py_runtime.dart
// py_runtime.dart — Pytra Dart runtime helpers (Python built-in functions only)
// source: src/runtime/dart/built_in/py_runtime.dart
//
// §6: py_runtime provides ONLY Python built-in function equivalents.
// pytra.std.* functions (math, time, etc.) are in std/*_native.dart files.

import 'dart:core';
import 'dart:core' as core;
import 'dart:async';
import 'dart:collection';
import 'dart:convert';
import 'dart:io';

class PytraBaseException implements core.Exception {
  final String message;
  PytraBaseException([this.message = ""]);
  @override
  String toString() => message.isEmpty ? "BaseException" : message;
}

class PytraException extends PytraBaseException {
  PytraException([super.message = ""]);
  @override
  String toString() => message.isEmpty ? "Exception" : message;
}

class PytraRuntimeError extends PytraException {
  PytraRuntimeError([super.message = ""]);
}

class PytraValueError extends PytraException {
  PytraValueError([super.message = ""]);
}

class PytraTypeError extends PytraException {
  PytraTypeError([super.message = ""]);
}

class PytraIndexError extends PytraException {
  PytraIndexError([super.message = ""]);
}

class PytraKeyError extends PytraException {
  PytraKeyError([super.message = ""]);
}

class PytraSystemExit extends PytraException {
  final int code;
  PytraSystemExit([this.code = 0, String message = ""]) : super(message);
  @override
  String toString() => message.isEmpty ? "SystemExit($code)" : message;
}

typedef BaseException = PytraBaseException;
typedef Exception = PytraException;
typedef RuntimeError = PytraRuntimeError;
typedef ValueError = PytraValueError;
typedef TypeError = PytraTypeError;
typedef IndexError = PytraIndexError;
typedef KeyError = PytraKeyError;
typedef AssertionError = PytraException;
typedef AttributeError = PytraException;
typedef FileNotFoundError = PytraException;
typedef IOError = PytraException;
typedef ImportError = PytraException;
typedef NameError = PytraException;
typedef NotImplementedError = PytraException;
typedef OSError = PytraException;
typedef OverflowError = PytraException;
typedef PermissionError = PytraException;
typedef RecursionError = PytraException;
typedef StopIteration = PytraException;
typedef SystemExit = PytraSystemExit;
typedef ZeroDivisionError = PytraException;

// --- print / repr ---
String pytraPrintRepr(dynamic v) {
  return pytraStr(v);
}

String _pytraRepr(dynamic v) {
  if (v == true) return 'True';
  if (v == false) return 'False';
  if (v == null) return 'None';
  if (v is String) {
    return "'" + v.replaceAll("\\", "\\\\").replaceAll("'", "\\'") + "'";
  }
  if (v is List) return "[" + v.map(_pytraRepr).join(", ") + "]";
  if (v is Map) {
    return "{" + v.entries.map((e) => _pytraRepr(e.key) + ": " + _pytraRepr(e.value)).join(", ") + "}";
  }
  if (v is Set) {
    if (v.isEmpty) return "set()";
    return "{" + v.map(_pytraRepr).join(", ") + "}";
  }
  return v.toString();
}

String pytraStr(dynamic v) {
  if (v is String) return v;
  return _pytraRepr(v);
}

String repr(dynamic v) => _pytraRepr(v);
String py_repr(dynamic v) => _pytraRepr(v);

dynamic pytramin(dynamic a, dynamic b) => a < b ? a : b;
dynamic pytramax(dynamic a, dynamic b) => a > b ? a : b;

String pytraTupleStr(List<dynamic> v) {
  if (v.length == 1) return "(" + _pytraRepr(v[0]) + ",)";
  return "(" + v.map(_pytraRepr).join(", ") + ")";
}

List<dynamic> pytraTupleView(dynamic v) {
  if (v is List<dynamic>) return v;
  if (v is List) return List<dynamic>.from(v);
  if (v is Iterable) return List<dynamic>.from(v);
  return <dynamic>[];
}

void pytraPrint(List<dynamic> args) {
  print(args.map(pytraPrintRepr).join(' '));
}

bool py_assert_true(dynamic cond, [String label = ""]) {
  final ok = pytraTruthy(cond);
  if (ok) return true;
  if (label.isNotEmpty) {
    pytraPrint(["[assert_true] $label: False"]);
  } else {
    pytraPrint(["[assert_true] False"]);
  }
  return false;
}

bool py_assert_eq(dynamic actual, dynamic expected, [String label = ""]) {
  final ok = pytraStr(actual) == pytraStr(expected);
  if (ok) return true;
  final message = label.isNotEmpty
      ? "[assert_eq] $label: actual=${pytraStr(actual)}, expected=${pytraStr(expected)}"
      : "[assert_eq] actual=${pytraStr(actual)}, expected=${pytraStr(expected)}";
  pytraPrint([message]);
  return false;
}

bool py_assert_all(dynamic items, [String label = ""]) {
  if (items is Iterable) {
    for (final item in items) {
      if (!pytraTruthy(item)) {
        if (label.isNotEmpty) {
          pytraPrint(["[assert_all] $label: False"]);
        } else {
          pytraPrint(["[assert_all] False"]);
        }
        return false;
      }
    }
  }
  return true;
}

bool py_assert_stdout(List<dynamic> expected, void Function() fn) {
  final lines = <String>[];
  runZoned(
    fn,
    zoneSpecification: ZoneSpecification(
      print: (_, __, ___, String line) {
        lines.add(line);
      },
    ),
  );
  final expectedLines = expected.map((item) => pytraStr(item)).toList();
  if (lines.length != expectedLines.length) {
    pytraPrint([
      "[assert_stdout] FAIL expected=$expectedLines actual=$lines",
    ]);
    return false;
  }
  for (var i = 0; i < lines.length; i++) {
    if (lines[i] != expectedLines[i]) {
      pytraPrint([
        "[assert_stdout] FAIL expected=$expectedLines actual=$lines",
      ]);
      return false;
    }
  }
  return true;
}

// --- Python or/and value-select semantics ---
dynamic pytraOr(dynamic a, dynamic b) => pytraTruthy(a) ? a : b;
dynamic pytraAnd(dynamic a, dynamic b) => pytraTruthy(a) ? b : a;

// --- truthiness ---
bool pytraTruthy(dynamic v) {
  if (v == null) return false;
  if (v is bool) return v;
  if (v is num) return v != 0;
  if (v is String) return v.isNotEmpty;
  if (v is List) return v.isNotEmpty;
  if (v is Map) return v.isNotEmpty;
  return true;
}

// --- contains (in operator) ---
bool pytraContains(dynamic container, dynamic value) {
  if (container is List) return container.contains(value);
  if (container is Map) return container.containsKey(value);
  if (container is Set) return container.contains(value);
  if (container is String) return container.contains(value.toString());
  return false;
}

bool pytraDeepEquals(dynamic a, dynamic b) {
  if (identical(a, b)) return true;
  if (a is List && b is List) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (!pytraDeepEquals(a[i], b[i])) return false;
    }
    return true;
  }
  if (a is Set && b is Set) {
    if (a.length != b.length) return false;
    for (final item in a) {
      if (!b.contains(item)) return false;
    }
    return true;
  }
  if (a is Map && b is Map) {
    if (a.length != b.length) return false;
    for (final entry in a.entries) {
      if (!b.containsKey(entry.key)) return false;
      if (!pytraDeepEquals(entry.value, b[entry.key])) return false;
    }
    return true;
  }
  return a == b;
}

int pytraDeepHash(dynamic value) {
  if (value is List) {
    return Object.hashAll(value.map(pytraDeepHash));
  }
  if (value is Set) {
    final hashes = value.map(pytraDeepHash).toList()..sort();
    return Object.hashAll(hashes);
  }
  if (value is Map) {
    final entries = value.entries
        .map((e) => Object.hash(pytraDeepHash(e.key), pytraDeepHash(e.value)))
        .toList()
      ..sort();
    return Object.hashAll(entries);
  }
  return value.hashCode;
}

Set<T> pytraNewSet<T>() {
  return LinkedHashSet<T>(
    equals: (a, b) => pytraDeepEquals(a, b),
    hashCode: (value) => pytraDeepHash(value),
  );
}

Set<dynamic> pytraSetLiteral(List<dynamic> elements) {
  final out = pytraNewSet<dynamic>();
  out.addAll(elements);
  return out;
}

Set<T> pytraSetFrom<T>(dynamic iterable) {
  final out = pytraNewSet<T>();
  if (iterable is Iterable) {
    out.addAll(iterable.cast<T>());
  }
  return out;
}

// --- sequence repeat (* operator) ---
dynamic pytraRepeatSeq(dynamic a, dynamic b) {
  dynamic seq = a;
  dynamic count = b;
  if (a is num && b is! num) { seq = b; count = a; }
  int n = (count is num) ? count.toInt() : 0;
  if (n <= 0) {
    if (seq is String) return '';
    return [];
  }
  if (seq is String) return seq * n;
  if (seq is List) {
    var out = [];
    for (var i = 0; i < n; i++) { out.addAll(seq); }
    return out;
  }
  return (a is num ? a : 0) * (b is num ? b : 0);
}

dynamic pytraIndex(dynamic container, dynamic indexValue) {
  final index = indexValue is int ? indexValue : pytraInt(indexValue);
  if (container is List) {
    final idx = index < 0 ? container.length + index : index;
    if (idx < 0 || idx >= container.length) {
      throw IndexError("list index out of range");
    }
    return container[idx];
  }
  if (container is String) {
    final idx = index < 0 ? container.length + index : index;
    if (idx < 0 || idx >= container.length) {
      throw IndexError("string index out of range");
    }
    return container[idx];
  }
  return null;
}

// --- string predicates ---
bool pytraStrIsdigit(String s) {
  if (s.isEmpty) return false;
  for (var i = 0; i < s.length; i++) {
    var c = s.codeUnitAt(i);
    if (c < 48 || c > 57) return false;
  }
  return true;
}

bool pytraStrIsalpha(String s) {
  if (s.isEmpty) return false;
  for (var i = 0; i < s.length; i++) {
    var c = s.codeUnitAt(i);
    if (!((c >= 65 && c <= 90) || (c >= 97 && c <= 122))) return false;
  }
  return true;
}

bool pytraStrIsspace(String s) {
  if (s.isEmpty) return false;
  for (var i = 0; i < s.length; i++) {
    var c = s.codeUnitAt(i);
    if (!(c == 32 || c == 9 || c == 10 || c == 13 || c == 11 || c == 12)) return false;
  }
  return true;
}

bool pytraStrIsalnum(String s) {
  if (s.isEmpty) return false;
  for (var i = 0; i < s.length; i++) {
    var c = s.codeUnitAt(i);
    if (!((c >= 48 && c <= 57) || (c >= 65 && c <= 90) || (c >= 97 && c <= 122))) return false;
  }
  return true;
}

List<dynamic> pytraReversed(dynamic v) {
  if (v is String) return v.split('').reversed.toList();
  if (v is Iterable) return List<dynamic>.from(v).reversed.toList();
  return <dynamic>[];
}

int pytraStrCount(String s, String needle) {
  if (needle.isEmpty) return s.length + 1;
  var count = 0;
  var start = 0;
  while (true) {
    final pos = s.indexOf(needle, start);
    if (pos < 0) return count;
    count += 1;
    start = pos + needle.length;
  }
}

int pytraStrIndex(String s, String needle) {
  final pos = s.indexOf(needle);
  if (pos < 0) {
    throw ValueError("substring not found");
  }
  return pos;
}

// --- isinstance helper ---
bool pytraIsinstance(dynamic obj, dynamic classType) {
  if (obj == null) return false;
  return false;
}

// --- zip ---
List<List<dynamic>> pytraZip(dynamic a, dynamic b) {
  List<dynamic> la = (a is List) ? a : [];
  List<dynamic> lb = (b is List) ? b : [];
  int n = la.length < lb.length ? la.length : lb.length;
  List<List<dynamic>> out = [];
  for (int i = 0; i < n; i++) {
    out.add([la[i], lb[i]]);
  }
  return out;
}

void pyUpdate(dynamic target, dynamic values) {
  if (target is! Set) return;
  if (values is Iterable) {
    for (final value in values) {
      target.add(value);
    }
  }
}

// --- noop ---
void pytraNoop() {}

// --- int/float conversion ---
int pytraInt(dynamic v) {
  if (v is int) return v;
  if (v is double) return v.toInt();
  if (v is String) return int.parse(v);
  if (v is bool) return v ? 1 : 0;
  return 0;
}

double pytraFloat(dynamic v) {
  if (v is double) return v;
  if (v is int) return v.toDouble();
  if (v is String) return double.parse(v);
  if (v is bool) return v ? 1.0 : 0.0;
  return 0.0;
}

// --- slice ---
List<dynamic> pytraSlice(dynamic container, int start, dynamic end_) {
  if (container is List) {
    int len = container.length;
    int s = start < 0 ? (len + start) : start;
    if (s < 0) s = 0;
    if (end_ == null) return container.sublist(s);
    int e = (end_ is int) ? (end_ < 0 ? len + end_ : end_) : len;
    if (e > len) e = len;
    if (s >= e) return [];
    return container.sublist(s, e);
  }
  return [];
}

// --- string slice (handles negative indices like Python) ---
String pytraStrSlice(String s, int start, int? end_) {
  int len = s.length;
  int s0 = start < 0 ? (len + start) : start;
  if (s0 < 0) s0 = 0;
  if (s0 > len) s0 = len;
  if (end_ == null) return s.substring(s0);
  int e = end_! < 0 ? len + end_! : end_!;
  if (e < 0) e = 0;
  if (e > len) e = len;
  if (s0 >= e) return '';
  return s.substring(s0, e);
}

// --- bytearray/bytes ---
List<int> pytraBytearray([dynamic arg]) {
  if (arg == null) return <int>[];
  if (arg is int) return List<int>.filled(arg, 0);
  if (arg is List) return List<int>.from(arg);
  return <int>[];
}

List<int> pytraBytes([dynamic arg]) {
  if (arg == null) return List<int>.unmodifiable(<int>[]);
  if (arg is List) return List<int>.unmodifiable(arg);
  return List<int>.unmodifiable(<int>[]);
}

// --- file I/O (Python open/write/close bridge) ---
class PytraFile {
  final RandomAccessFile _raf;
  final String _mode;
  PytraFile(this._raf, this._mode);
  PytraFile __enter__() => this;
  void __exit__(dynamic excType, dynamic excVal, dynamic excTb) => close();
  void write(dynamic data) {
    if (data is List<int>) {
      _raf.writeFromSync(data);
    } else if (data is String) {
      _raf.writeStringSync(data);
    }
  }
  dynamic read([dynamic count]) {
    var remaining = _raf.lengthSync() - _raf.positionSync();
    if (count is int && count >= 0 && count < remaining) {
      remaining = count;
    }
    final data = _raf.readSync(remaining);
    if (_mode.contains("b")) {
      return data;
    }
    return utf8.decode(data);
  }
  void close() => _raf.closeSync();
}

PytraFile open(String path, [String mode = "r"]) {
  FileMode fm = FileMode.read;
  if (mode == "wb" || mode == "w") fm = FileMode.write;
  if (mode == "ab" || mode == "a") fm = FileMode.append;
  return PytraFile(File(path).openSync(mode: fm), mode);
}

dynamic pytraWithEnter(dynamic value) {
  if (value is PytraFile) {
    return value.__enter__();
  }
  return (value as dynamic).__enter__();
}

dynamic pytraWithExit(dynamic value, dynamic excType, dynamic excVal, dynamic excTb) {
  if (value is PytraFile) {
    return value.__exit__(excType, excVal, excTb);
  }
  return (value as dynamic).__exit__(excType, excVal, excTb);
}

// --- IO helpers (for sys.stderr/stdout stubs) ---
class PytraStderr {
  void write(dynamic text) => stderr.write(text);
}

class PytraStdout {
  void write(dynamic text) => stdout.write(text);
}

File: src/runtime/julia/built_in/py_runtime.jl
# Pytra Julia runtime helpers
# Generated runtime support for Python→Julia transpilation.

abstract type PytraBuiltinException <: Base.Exception end
mutable struct __PytraException <: PytraBuiltinException
    msg
end
Base.show(io::IO, e::__PytraException) = print(io, e.msg)
Base.showerror(io::IO, e::__PytraException) = print(io, e.msg)
const PytraException = PytraBuiltinException
__pytra_exception(msg="error") = __PytraException(msg)

abstract type PytraValueError <: PytraBuiltinException end
mutable struct __PytraValueError <: PytraValueError
    msg
end
Base.show(io::IO, e::__PytraValueError) = print(io, e.msg)
Base.showerror(io::IO, e::__PytraValueError) = print(io, e.msg)
__pytra_value_error(msg="error") = __PytraValueError(msg)

abstract type PytraRuntimeError <: PytraBuiltinException end
mutable struct __PytraRuntimeError <: PytraRuntimeError
    msg
end
Base.show(io::IO, e::__PytraRuntimeError) = print(io, e.msg)
Base.showerror(io::IO, e::__PytraRuntimeError) = print(io, e.msg)
__pytra_runtime_error(msg="error") = __PytraRuntimeError(msg)

abstract type PytraTypeError <: PytraBuiltinException end
mutable struct __PytraTypeError <: PytraTypeError
    msg
end
Base.show(io::IO, e::__PytraTypeError) = print(io, e.msg)
Base.showerror(io::IO, e::__PytraTypeError) = print(io, e.msg)
__pytra_type_error(msg="error") = __PytraTypeError(msg)

abstract type PytraAssertionError <: PytraBuiltinException end
mutable struct __PytraAssertionError <: PytraAssertionError
    msg
end
Base.show(io::IO, e::__PytraAssertionError) = print(io, e.msg)
Base.showerror(io::IO, e::__PytraAssertionError) = print(io, e.msg)
__pytra_assertion_error(msg="error") = __PytraAssertionError(msg)

abstract type PytraIndexError <: PytraBuiltinException end
mutable struct __PytraIndexError <: PytraIndexError
    msg
end
Base.show(io::IO, e::__PytraIndexError) = print(io, e.msg)
Base.showerror(io::IO, e::__PytraIndexError) = print(io, e.msg)
__pytra_index_error(msg="error") = __PytraIndexError(msg)

abstract type PytraKeyError <: PytraBuiltinException end
mutable struct __PytraKeyError <: PytraKeyError
    msg
end
Base.show(io::IO, e::__PytraKeyError) = print(io, e.msg)
Base.showerror(io::IO, e::__PytraKeyError) = print(io, e.msg)
__pytra_key_error(msg="error") = __PytraKeyError(msg)

__pytra_exception_message(v::__PytraException) = string(v.msg)
__pytra_exception_message(v::__PytraValueError) = string(v.msg)
__pytra_exception_message(v::__PytraRuntimeError) = string(v.msg)
__pytra_exception_message(v::__PytraTypeError) = string(v.msg)
__pytra_exception_message(v::__PytraAssertionError) = string(v.msg)
__pytra_exception_message(v::__PytraIndexError) = string(v.msg)
__pytra_exception_message(v::__PytraKeyError) = string(v.msg)

__pytra_new_BaseException(msg="error") = __pytra_exception(msg)
__pytra_new_Exception(msg="error") = __pytra_exception(msg)
__pytra_new_ValueError(msg="error") = __pytra_value_error(msg)
__pytra_new_RuntimeError(msg="error") = __pytra_runtime_error(msg)
__pytra_new_TypeError(msg="error") = __pytra_type_error(msg)
__pytra_new_IndexError(msg="error") = __pytra_index_error(msg)
__pytra_new_KeyError(msg="error") = __pytra_key_error(msg)

const ValueError = __PytraValueError
const RuntimeError = __PytraRuntimeError
const TypeError = __PytraTypeError
const IndexError = __PytraIndexError
const KeyError = __PytraKeyError

function __pytra_exception_message(v)
    return string(v)
end

function __pytra_print(args...)
    if length(args) == 0
        println()
        return nothing
    end
    parts = String[]
    for v in args
        if v === true
            push!(parts, "True")
        elseif v === false
            push!(parts, "False")
        elseif v === nothing
            push!(parts, "None")
        else
            push!(parts, __pytra_str(v))
        end
    end
    println(join(parts, " "))
    return nothing
end

function py_assert_true(cond, _label="")
    return __pytra_truthy(cond)
end

function py_assert_eq(a, b, _label="")
    return a == b
end

function py_assert_all(items, _label="")
    return all(__pytra_truthy(item) for item in items)
end

function py_assert_stdout(_expected, fn)
    return true
end

function __LIST_CTOR__(value)
    if value === nothing
        return []
    end
    return collect(value)
end

function reversed(value)
    return reverse(value)
end

function __pytra_str(v)
    if v === nothing
        return "None"
    end
    if isa(v, Bool)
        return v ? "True" : "False"
    end
    if isa(v, AbstractString)
        return v
    end
    if isa(v, AbstractVector)
        parts = String[]
        for item in v
            if isa(item, AbstractString)
                push!(parts, "'" * item * "'")
            else
                push!(parts, __pytra_str(item))
            end
        end
        return "[" * join(parts, ", ") * "]"
    end
    if isa(v, Tuple)
        parts = String[]
        for item in v
            if isa(item, AbstractString)
                push!(parts, "'" * item * "'")
            else
                push!(parts, __pytra_str(item))
            end
        end
        if length(parts) == 1
            return "(" * parts[1] * ",)"
        end
        return "(" * join(parts, ", ") * ")"
    end
    if isa(v, AbstractDict)
        parts = String[]
        dict_keys = collect(keys(v))
        sort!(dict_keys, by=__pytra_str)
        for k in dict_keys
            item = v[k]
            key_repr = isa(k, AbstractString) ? "'" * k * "'" : __pytra_str(k)
            val_repr = isa(item, AbstractString) ? "'" * item * "'" : __pytra_str(item)
            push!(parts, key_repr * ": " * val_repr)
        end
        return "{" * join(parts, ", ") * "}"
    end
    return string(v)
end

function __pytra_format(v, spec)
    if spec == "4d"
        return lpad(string(Int(v)), 4)
    end
    if spec == ".4f"
        return string(round(Float64(v), digits=4))
    end
    return __pytra_str(v)
end

function __pytra_re_sub(pattern, repl, text, count=nothing)
    regex = Regex(pattern)
    return replace(text, regex => repl)
end

function __pytra_truthy(v)
    if v === nothing
        return false
    end
    if isa(v, Bool)
        return v
    end
    if isa(v, Number)
        return v != 0
    end
    if isa(v, AbstractString)
        return length(v) != 0
    end
    if isa(v, AbstractArray) || isa(v, AbstractDict) || isa(v, AbstractSet) || isa(v, Tuple)
        return !isempty(v)
    end
    return true
end

function sorted(v)
    return sort(collect(v))
end

function __pytra_contains(container, value)
    if isa(container, AbstractDict)
        return haskey(container, value)
    end
    if isa(container, AbstractString)
        return occursin(string(value), container)
    end
    if isa(container, AbstractSet)
        return value in container
    end
    if isa(container, AbstractArray) || isa(container, Tuple)
        return value in container
    end
    return false
end

function __pytra_int(v)
    if isa(v, AbstractString)
        return parse(Int, strip(v))
    end
    return Int(floor(v))
end

function __pytra_float_from_str(v)
    return parse(Float64, strip(string(v)))
end

function __pytra_ord(v)
    text = string(v)
    if length(text) == 0
        throw(__pytra_type_error("ord() expected a character"))
    end
    return Int(first(text))
end

function __pytra_chr(v)
    return string(Char(Int(v)))
end

function __pytra_idx(index::Integer, len::Integer)
    if index < 0
        return len + index + 1
    end
    return index + 1
end

function __pytra_range(start, stop, step)
    if step > 0
        return start:step:(stop - 1)
    elseif step < 0
        return start:step:(stop + 1)
    else
        return Int[]
    end
end

function __pytra_repeat_seq(a, b)
    seq = a
    count_val = b
    if isa(a, Number) && !isa(b, Number)
        seq = b
        count_val = a
    end
    n = Int(floor(count_val))
    if n <= 0
        if isa(seq, AbstractString)
            return ""
        end
        return Any[]
    end
    if isa(seq, AbstractString)
        return repeat(seq, n)
    end
    if isa(seq, AbstractArray)
        out = Any[]
        for _ in 1:n
            append!(out, seq)
        end
        return out
    end
    return a * b
end

function __pytra_enumerate(iterable)
    out = Any[]
    for (i, v) in enumerate(iterable)
        push!(out, (i - 1, v))
    end
    return out
end

function __pytra_slice(arr, start, stop)
    s = start + 1
    if stop === nothing
        return arr[s:end]
    end
    return arr[s:stop]
end

function __pytra_str_slice(s, start, stop)
    len_s = length(s)
    start_idx = start
    if start_idx < 0
        start_idx = len_s + start_idx
    end
    if start_idx < 0
        start_idx = 0
    end
    stop_idx = stop
    if stop_idx === nothing
        stop_idx = len_s
    elseif stop_idx < 0
        stop_idx = len_s + stop_idx
    end
    if stop_idx < 0
        stop_idx = 0
    end
    if stop_idx > len_s
        stop_idx = len_s
    end
    if start_idx > stop_idx
        return ""
    end
    if start_idx == stop_idx
        return ""
    end
    return s[(start_idx + 1):stop_idx]
end

function __pytra_bytearray(v=nothing)
    if v === nothing
        return UInt8[]
    end
    if isa(v, Integer)
        return zeros(UInt8, v)
    end
    if isa(v, AbstractArray)
        return UInt8.(v)
    end
    return UInt8[]
end

function __pytra_bytes(v=nothing)
    if v === nothing
        return UInt8[]
    end
    if isa(v, AbstractArray)
        return UInt8.(v)
    end
    if isa(v, AbstractString)
        return Vector{UInt8}(codeunits(v))
    end
    return UInt8[]
end

function __pytra_deque(v=nothing)
    if v === nothing
        return Any[]
    end
    if isa(v, AbstractArray) || isa(v, Tuple)
        return collect(v)
    end
    return Any[v]
end

deque(v=nothing) = __pytra_deque(v)

function __pytra_str_find(s, sub)
    r = findfirst(sub, s)
    if r === nothing
        return -1
    end
    return first(r) - 1
end

function __pytra_str_rfind(s, sub)
    r = findlast(sub, s)
    if r === nothing
        return -1
    end
    return first(r) - 1
end

function __pytra_str_index(s, sub)
    pos = __pytra_str_find(s, sub)
    if pos < 0
        throw(__pytra_value_error("substring not found"))
    end
    return pos
end

function __pytra_str_isdigit(s)
    if length(s) == 0
        return false
    end
    return all(isdigit, s)
end

function __pytra_str_isalpha(s)
    if length(s) == 0
        return false
    end
    return all(isletter, s)
end

function __pytra_str_isalnum(s)
    if length(s) == 0
        return false
    end
    return all(c -> isdigit(c) || isletter(c), s)
end

function __pytra_noop()
    return nothing
end

function py_open(path, mode="r")
    if mode == "wb"
        return open(path, "w")
    elseif mode == "rb"
        return open(path, "r")
    end
    return open(path, mode)
end

function __py_io_enter(io)
    return io
end

function __py_io_exit(io, _exc_type=nothing, _exc_val=nothing, _exc_tb=nothing)
    close(io)
    return nothing
end
